engine/src/main/java/org/pentaho/di/trans/steps/jobexecutor/JobExecutorMeta.java

```diff
@@ -0,0 +1,1483 @@
+    String realTransname = transMeta.environmentSubstitute( jobName );
+    ResourceReference reference = new ResourceReference( stepInfo );
+    references.add( reference );
+
+    if ( !Utils.isEmpty( realFilename ) ) {
+      // Add the filename to the references, including a reference to this step
+      // meta data.
+      //
+      reference.getEntries().add( new ResourceEntry( realFilename, ResourceType.ACTIONFILE ) );
+    } else if ( !Utils.isEmpty( realTransname ) ) {
+      // Add the filename to the references, including a reference to this step
+      // meta data.
+      //
+      reference.getEntries().add( new ResourceEntry( realTransname, ResourceType.ACTIONFILE ) );
+      references.add( reference );
+    }
+    return references;
+  }
+
+  /**
+   * This method was created exclusively for tests
+   * to bypass the mock static final method
+   * without using PowerMock
+   *
+   * @param executorMeta
+   * @param rep
+   * @param space
+   * @return JobMeta
+   * @throws KettleException
+   */
+  JobMeta loadJobMetaProxy( JobExecutorMeta executorMeta, Repository rep,
+                           VariableSpace space ) throws KettleException {
+    return loadJobMeta( executorMeta, rep, space );
+  }
+
+  @Override
+  public String exportResources( VariableSpace space, Map<String, ResourceDefinition> definitions,
+    ResourceNamingInterface resourceNamingInterface, Repository repository, IMetaStore metaStore ) throws KettleException {
+    try {
+      // Try to load the transformation from repository or file.
+      // Modify this recursively too...
+      //
+      // NOTE: there is no need to clone this step because the caller is
+      // responsible for this.
+      //
+      // First load the executor job metadata...
+      //
+      JobMeta executorJobMeta = loadJobMetaProxy( this, repository, space );
+
+      // Also go down into the mapping transformation and export the files
+      // there. (mapping recursively down)
+      //
+      String proposedNewFilename =
+        executorJobMeta.exportResources(
+          executorJobMeta, definitions, resourceNamingInterface, repository, metaStore );
+
+      // To get a relative path to it, we inject
+      // ${Internal.Entry.Current.Directory}
+      //
+      String newFilename =
+        "${" + Const.INTERNAL_VARIABLE_ENTRY_CURRENT_DIRECTORY + "}/" + proposedNewFilename;
+
+      // Set the correct filename inside the XML.
+      //
+      executorJobMeta.setFilename( newFilename );
+
+      // exports always reside in the root directory, in case we want to turn
+      // this into a file repository...
+      //
+      executorJobMeta.setRepositoryDirectory( new RepositoryDirectory() );
+
+      // change it in the job entry
+      //
+      fileName = newFilename;
+
+      setSpecificationMethod( ObjectLocationSpecificationMethod.FILENAME );
+
+      return proposedNewFilename;
+    } catch ( Exception e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)